### PR TITLE
Move to golang 1.18 and later

### PR DESCRIPTION
**Pull request by rhatdan**
Github.com is reporting security issues on older versions of golang.

### CODE REVIEW

**Comment by rhatdan** [comment]
@nalind @saschagrunert @flouthoc @giuseppe @vrothberg @umohnani8 PTAL

**Review by saschagrunert — APPROVED** [review]
LGTM :+1:

**Comment by nalind** [comment]
This modifies many files under `tests/tools/vendor`, but not `tests/tools/vendor/modules.txt`.  What's going on there?

File: pkg/archive/copy_unix_test.go [hunk 1]
```diff
@@ -256,8 +256,9 @@ func TestCopyErrDstNotDir(t *testing.T) {
 //
 
 // A. SRC specifies a file and DST (no trailing path separator) doesn't
-//    exist. This should create a file with the name DST and copy the
-//    contents of the source file into it.
+//
```
[the remark this answers is not in the document]
Comment (rhatdan) [reply pkg/archive/copy_unix_test.go:259]: This patch was done with:

$ find . -name \*go -exec gofmt -s -w {} \; -print

Followed by make vendor-in-container.

**Comment by rhatdan** [comment]
@cevich Any idea on the vendor failure. Is this a matter of updating the golang inside of the VMs?

**Comment by cevich** [comment]
That's strange...the fedora 37 VMs should be on 1.19...looking...

**Comment by cevich** [comment]
...oh! This task is running in a container, that's where the version comes from.  In `.cirrus.yml`:

```yaml
vendor_task:
    container:
        image: golang:1.17
    modules_cache:
        fingerprint_script: cat go.sum
        folder: $GOPATH/pkg/mod
    build_script: make vendor
    test_script: hack/tree_status.sh
```

That's confusing.  It means multiple files need to be updated to reflect a new version in CI.  Ugg.  Let me see if I can make this task follow the latest Fedora VM instead...

**Comment by cevich** [comment]
...nvm, it seems you know what you're doing.

File: pkg/archive/copy_unix_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nalind) [reply pkg/archive/copy_unix_test.go:259]: > This patch was done with:
> 
> $ find . -name *go -exec gofmt -s -w {} ; -print

Okay, that would also modify sources in vendored dependencies, which we don't want to be doing.
 
> Followed by make vendor-in-container.

The target doesn't touch vendored dependencies under `tests/tools`, which I think is what this warning notification was focusing on.  Maybe `go get golang.org/x/sys && go mod vendor && go mod tidy` from there?

**Comment by nalind** [comment]
This may also require a bump to the version of the linter installed in `tests/tools/Makefile`.

**Comment by rhatdan** [comment]
@nalind Any idea what is going wrong now?

**Comment by nalind** [comment]
When I ran it locally, "mount-image" (image-mount.bats:17) failed because `btrfs driver does not support mount options` (it won't say which one), and "container-on-mounted-image" (image-mount.bats:79) failed because it tried to create an image with the same name as one it has just created (a copy/paste error?).
The other "mount-image" (mount.bats:51) fails because the CLI test helper understands "-r" and "--ro", but not "--read-only" (either the test or the helper should be updated).

**Comment by rhatdan** [comment]
Thanks for pointing to those files, which I have no idea where they came from.

File: pkg/archive/archive_linux_test.go [hunk 2]
```diff
@@ -21,7 +21,8 @@ import (
 // ├── d2     # opaque, 0750
 // │   └── f1 # empty file, 0660
 // └── d3     # 0700
-//     └── f1 # whiteout, 0000
+//
```
Comment (nalind) [inline pkg/archive/archive_linux_test.go:24]: Extra blank line inserted here for some reason.

File: pkg/archive/copy_unix_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nalind) [reply pkg/archive/copy_unix_test.go:259]: This has not been resolved.

File: pkg/chunked/cache_linux.go [hunk 3]
```diff
@@ -398,7 +398,8 @@ func prepareMetadata(manifest []byte) ([]*internal.FileMetadata, error) {
 	toc, err := unmarshalToc(manifest)
 	if err != nil {
 		// ignore errors here.  They might be caused by a different manifest format.
-		return nil, nil
+		logrus.Debugf("could not unmasshal manifest: %v", err)
```
Comment (nalind) [inline pkg/chunked/cache_linux.go:401]: Typo: "unmasshal" should be "unmarshal".

File: pkg/parsers/parsers.go [hunk 4]
```diff
@@ -24,13 +24,14 @@ func ParseKeyValueOpt(opt string) (string, string, error) {
...
 // set to `true`.
 // Supported formats:
-//     7
-//     1-6
-//     0,3-4,7,8-10
-//     0-0,0,1-7
-//     03,1-3      <- this is gonna get parsed as [1,2,3]
-//     3,2,1
-//     0-2,3,1
+//
```
Comment (nalind) [inline pkg/parsers/parsers.go:27]: Extra blank line inserted.

File: pkg/chunked/compression_linux.go [hunk 5]
```diff
@@ -149,7 +150,7 @@ func readEstargzChunkedManifest(blobStream ImageSourceSeekable, blobSize int64,
 // readZstdChunkedManifest reads the zstd:chunked manifest from the seekable stream blobStream.  The blob total size must
 // be specified.
 // This function uses the io.github.containers.zstd-chunked. annotations when specified.
-func readZstdChunkedManifest(blobStream ImageSourceSeekable, blobSize int64, annotations map[string]string) ([]byte, int64, error) {
+func readZstdChunkedManifest(ctx context.Context, blobStream ImageSourceSeekable, blobSize int64, annotations map[string]string) ([]byte, int64, error) {
```
[the remark this answers is not in the document]
Comment (rhatdan) [reply pkg/chunked/compression_linux.go:153]: There was a linter that was complaining about it, so I added it and then had to nolint below.  I can remove this, but I think we should start to use more of the linters for storage.

File: pkg/parsers/parsers.go (hunk 4 above)
Comment (rhatdan) [reply pkg/parsers/parsers.go:27]: I think go fmt is doing this.

File: pkg/chunked/cache_linux.go (hunk 3 above)
Comment (rhatdan) [reply pkg/chunked/cache_linux.go:401]: Fixed

File: pkg/archive/archive_linux_test.go (hunk 2 above)
Comment (rhatdan) [reply pkg/archive/archive_linux_test.go:24]: GO fmt is doing this I belive.

File: pkg/chunked/compression_linux.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (rhatdan) [reply pkg/chunked/compression_linux.go:153]: +       decoder, err := zstd.NewReader(nil) //nolint:contextcheck

[the remark this answers is not in the document]
Comment (rhatdan) [reply pkg/chunked/compression_linux.go:153]: The context check wants it passed in here, but I did not know the code well enough to add it. @giuseppe PTAL

**Review by giuseppe — APPROVED** [review]
LGTM

**Pull request merged**